src/bun.js/bindings/NodeVMModule.cpp

```diff
@@ -0,0 +1,395 @@
+}
+
+JSC_DEFINE_HOST_FUNCTION(jsNodeVmModuleGetStatusCode, (JSC::JSGlobalObject * globalObject, JSC::CallFrame* callFrame))
+{
+    auto* thisObject = jsCast<NodeVMSourceTextModule*>(callFrame->thisValue());
+    return JSValue::encode(JSC::jsNumber(static_cast<uint32_t>(thisObject->status())));
+}
+
+JSC_DEFINE_HOST_FUNCTION(jsNodeVmModuleGetStatus, (JSC::JSGlobalObject * globalObject, JSC::CallFrame* callFrame))
+{
+    auto* thisObject = jsCast<NodeVMSourceTextModule*>(callFrame->thisValue());
+
+    using enum NodeVMModule::Status;
+    switch (thisObject->status()) {
+    case Unlinked:
+        return JSValue::encode(JSC::jsString(globalObject->vm(), WTF::String("unlinked"_s)));
+    case Linking:
+        return JSValue::encode(JSC::jsString(globalObject->vm(), WTF::String("linking"_s)));
+    case Linked:
+        return JSValue::encode(JSC::jsString(globalObject->vm(), WTF::String("linked"_s)));
+    case Evaluating:
+        return JSValue::encode(JSC::jsString(globalObject->vm(), WTF::String("evaluating"_s)));
+    case Evaluated:
+        return JSValue::encode(JSC::jsString(globalObject->vm(), WTF::String("evaluated"_s)));
+    case Errored:
+        return JSValue::encode(JSC::jsString(globalObject->vm(), WTF::String("errored"_s)));
+    default:
+        return JSC::encodedJSUndefined();
+    }
+}
+
+JSC_DEFINE_HOST_FUNCTION(jsNodeVmModuleGetNamespace, (JSC::JSGlobalObject * globalObject, JSC::CallFrame* callFrame))
+{
+    auto* thisObject = jsCast<NodeVMSourceTextModule*>(callFrame->thisValue());
+    return JSValue::encode(thisObject->namespaceObject(globalObject));
+}
+
+JSC_DEFINE_HOST_FUNCTION(jsNodeVmModuleGetError, (JSC::JSGlobalObject * globalObject, JSC::CallFrame* callFrame))
+{
+    VM& vm = globalObject->vm();
+    auto scope = DECLARE_THROW_SCOPE(vm);
+
+    if (auto* thisObject = jsCast<NodeVMSourceTextModule*>(callFrame->thisValue())) {
+        if (JSC::Exception* exception = thisObject->evaluationException()) {
+            return JSValue::encode(exception->value());
+        }
+        throwError(globalObject, scope, ErrorCode::ERR_VM_MODULE_STATUS, "Module status must be errored"_s);
+        return {};
+    }
+
+    throwTypeError(globalObject, scope, "This function must be called on a SourceTextModule or SyntheticModule"_s);
+    return {};
+}
+
+JSC_DEFINE_HOST_FUNCTION(jsNodeVmModuleGetModuleRequests, (JSC::JSGlobalObject * globalObject, JSC::CallFrame* callFrame))
+{
+    auto* thisObject = jsCast<NodeVMModule*>(callFrame->thisValue());
+
+    if (auto* sourceTextModule = jsDynamicCast<NodeVMSourceTextModule*>(callFrame->thisValue())) {
+        sourceTextModule->ensureModuleRecord(globalObject);
+    }
+
+    const WTF::Vector<NodeVMModuleRequest>& requests = thisObject->moduleRequests();
+
+    JSArray* array = constructEmptyArray(globalObject, nullptr, requests.size());
+
+    for (unsigned i = 0; const NodeVMModuleRequest& request : requests) {
```

Comment:
I suggest constructing the array at the end of this, appending to a MarkedArgumentsBuffer, checking for exceptions and then passing the MarkedArgumentsBuffer to constructEmptyArray.

```suggestion
    MarkedArgumentsBuffer array;
    for ( const NodeVMModuleRequest& request : requests) {
          array.append(request.toJS(globalObject));
          RETURN_IF_EXCEPTION(scope, {});
     }
     return JSValue::encode(constructEmptyArray(globalObject, array));
```